src/ssl.c

```diff
@@ -54563,65 +54563,96 @@ int wolfSSL_BN_rand(WOLFSSL_BIGNUM* bn, int bits, int top, int bottom)
+            len++;
+        }
 
-    /* has to be a length of at least 1 since we set buf[0] and buf[len-1] */
-    if (len < 1) {
-        return WOLFSSL_FAILURE;
+    #ifdef WOLFSSL_SMALL_STACK
+        buff   = (byte*)XMALLOC(1024,        NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        tmpRNG = (WC_RNG*) XMALLOC(sizeof(WC_RNG), NULL, DYNAMIC_TYPE_RNG);
+        if (buff == NULL || tmpRNG == NULL) {
+            XFREE(buff,   NULL, DYNAMIC_TYPE_TMP_BUFFER);
+            XFREE(tmpRNG, NULL, DYNAMIC_TYPE_RNG);
+            ret = WOLFSSL_FAILURE;
+        }
+    #endif
     }
 
-#ifdef WOLFSSL_SMALL_STACK
-    buff   = (byte*)XMALLOC(1024,        NULL, DYNAMIC_TYPE_TMP_BUFFER);
-    tmpRNG = (WC_RNG*) XMALLOC(sizeof(WC_RNG), NULL, DYNAMIC_TYPE_RNG);
-    if (buff == NULL || tmpRNG == NULL) {
-        XFREE(buff,   NULL, DYNAMIC_TYPE_TMP_BUFFER);
-        XFREE(tmpRNG, NULL, DYNAMIC_TYPE_RNG);
-        return ret;
+    if (ret == WOLFSSL_SUCCESS) {
+        if (wc_InitRng(tmpRNG) == 0) {
+            rng = tmpRNG;
+            initTmpRng = 1;
+        }
+        else if (initGlobalRNG) {
+            rng = &globalRNG;
+        }
+        else {
+            WOLFSSL_MSG("Failed to get an RNG");
+            ret = WOLFSSL_FAILURE;
+        }
     }
-#endif
 
-    if (bn == NULL || bn->internal == NULL)
-        WOLFSSL_MSG("Bad function arguments");
-    else if (wc_InitRng(tmpRNG) == 0) {
-        rng = tmpRNG;
-        initTmpRng = 1;
+    if (ret == WOLFSSL_SUCCESS && wc_RNG_GenerateBlock(rng, buff, len) != 0) {
+        WOLFSSL_MSG("wc_RNG_GenerateBlock failed");
+        ret = WOLFSSL_FAILURE;
     }
-    else if (initGlobalRNG)
-        rng = &globalRNG;
-
-    if (rng) {
-        if (wc_RNG_GenerateBlock(rng, buff, len) != 0)
-            WOLFSSL_MSG("Bad wc_RNG_GenerateBlock");
-        else {
-            buff[0]     |= 0x80 | 0x40;
-            buff[len-1] |= 0x01;
-
-            if (mp_read_unsigned_bin((mp_int*)bn->internal,buff,len) != MP_OKAY)
-                WOLFSSL_MSG("mp read bin failed");
-            else
-                ret = WOLFSSL_SUCCESS;
+    if (ret == WOLFSSL_SUCCESS &&
+        mp_read_unsigned_bin((mp_int*)bn->internal,buff,len) != MP_OKAY) {
+            WOLFSSL_MSG("mp_read_unsigned_bin failed");
+            ret = WOLFSSL_FAILURE;
+    }
+    if (ret == WOLFSSL_SUCCESS) {
+        /* Truncate to requested bit length. */
+        mp_rshb((mp_int*)bn->internal, 8 - lastBits);
+        if (bottom && mp_set_bit((mp_int*)bn->internal, 0) != MP_OKAY) {
+            WOLFSSL_MSG("Failed to set bit 0");
+            ret = WOLFSSL_FAILURE;
         }
     }
+    if (ret == WOLFSSL_SUCCESS) {
+        if (top == 0) {
+            if (mp_set_bit((mp_int*)bn->internal, bits - 1) != MP_OKAY) {
+                WOLFSSL_MSG("Failed to set top bit");
+                ret = WOLFSSL_FAILURE;
+            }
+        }
+        else if (top > 0) {
+            if (mp_set_bit((mp_int*)bn->internal, bits - 1) != MP_OKAY || 
```

Comment:
Not that I can think of!
Just easier to see what is happening.

Reply:
I kind of see it the other way. I think it's clearer if we handle `top` and `bottom` at the `mp_int` level rather than handling `top` with the generated data and `bottom` with the `mp_int`.